[MNT] Upgrade code style beyond python 3.8

yarnabrina opened the pull request:
Towards #5791

Depends on #6329

yarnabrina commented:
@fkiraly @benHeid any idea what change is causing the TypeError from subclass check? I can't find a change that will affect this type of logic.

yarnabrina commented:
@fkiraly I've rebased this branch with `main` and removed `Callable` changes. I'm sort of convinced that those are failing because of the linked `skbase` code is doing something wrong, but haven't figured that out. To keep it simple and quick for release, we can delay it post release.

fkiraly commented:
Could you kindly be more precise @yarnabrina?

If I follow the link, I do not see a subclass check. Could you explain please, with links, which line (link please) is causing which failure (link please), and what is the lowest-level reason, e.g., exception raised because (reason)?

yarnabrina commented:
This is the `issubclass` check: https://github.com/sktime/skbase/blob/068367e78bc928b421ddf0978d593287475c556d/skbase/lookup/_lookup.py#L166

This is called [here](https://github.com/sktime/skbase/blob/068367e78bc928b421ddf0978d593287475c556d/skbase/lookup/_lookup.py#L880), when `all_estimators` is called in `sktime` through `two `all_objects` in `sktime` and `skbase`.

This check is failing when `klass` becomes something like `collections.abc.Iterator[tuple[numpy.ndarray, numpy.ndarray]]`.

yarnabrina commented on sktime/split/base/_base_splitter.py:
```diff
@@ -4,7 +4,8 @@
 __author__ = ["fkiraly", "khrapovs", "mateuja", "mloning"]
 
-from typing import Iterator, Optional, Tuple
+from collections.abc import Iterator
```
```suggestion
from typing import Iterator
```

place 1

yarnabrina commented on sktime/split/base/_common.py:
```diff
@@ -2,7 +2,8 @@
 # copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
 
-from typing import Iterator, Optional, Tuple, Union
+from collections.abc import Iterator
```
```suggestion
from typing import Iterator
```

place 2

yarnabrina reviewed (commented):
Undoing these two changes are enough to make this particular CI job pass (verified locally). But this is not compatible with `pyupgrade --py39-plus` and code quality check will fail.

fkiraly commented:
> Undoing these two changes are enough to make this particular CI job pass (verified locally). But this is not compatible with `pyupgrade --py39-plus` and code quality check will fail.

What do you mean by "undoing"?

> This check is failing when

Which check?

Could you explain please, with links, which line (✔️) is causing which failure (link please), and what is the lowest-level reason, e.g., exception raised because (reason)?

fkiraly commented:
ok, I get it now. Could you kindly explain, @yarnabrina what the mechanism is that is leading to the failure? I.e., where and how does the interpreter infer form that it needs to check for the arg being a class?

yarnabrina commented:
Please see [this](https://github.com/sktime/sktime/pull/6330#issuecomment-2143940244) comment above.

In my understanding, this PR does not introduce any new calls to `skbase` that was not there before, it's just that the `klass` inference during `all_estimators` calls seem to have changed behaviour and that's where I suspect the problem is. _I have not been able to pinpoint the exact problem._

I added some prints in `_filter_by_class` to print `klass` and `class_filter`. This is the comparison between the prints with `typing.Iterator` and `collections.abc.Iterator`.

![image](https://github.com/sktime/sktime/assets/39331844/5a0ea09e-4ebd-46c4-9526-dc931b8a182e)

I do not know how you extract `klass` and what exactly it is supposed to contain, but I've noticed that all the failures seem to be related to `BaseEarlyClassifier` and `BaseDeepNetwork`.

If it helps you in navigation, this is the actual traceback.

```
=================================== FAILURES ===================================
___________ test_all_estimators_by_scitype[estimator_scitype20-True] ___________

estimator_scitype = ['aligner', 'early_classifier'], return_names = True

    @pytest.mark.parametrize("return_names", [True, False])
    @pytest.mark.parametrize("estimator_scitype", estimator_scitype_fixture)
    def test_all_estimators_by_scitype(estimator_scitype, return_names):
        """Check that all_estimators return argument has correct type."""
>       estimators = all_estimators(
            estimator_types=estimator_scitype,
            return_names=return_names,
        )

sktime/registry/tests/test_lookup.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sktime/registry/_lookup.py:195: in all_estimators
    result = all_objects(
/path/to/env/lib/python3.10/site-packages/skbase/lookup/_lookup.py:877: in all_objects
    all_estimators = [
/path/to/env/lib/python3.10/site-packages/skbase/lookup/_lookup.py:878: in <listcomp>
    (n, est) for (n, est) in all_estimators if _filter_by_class(est, obj_types)
/path/to/env/lib/python3.10/site-packages/skbase/lookup/_lookup.py:165: in _filter_by_class
    return issubclass(klass, class_filter)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'sktime.classification.early_classification.base.BaseEarlyClassifier'>
subclass = collections.abc.Iterator[tuple[numpy.ndarray, numpy.ndarray]]

    def __subclasscheck__(cls, subclass):
        """Override for issubclass(subclass, cls)."""
>       return _abc_subclasscheck(cls, subclass)
E       TypeError: issubclass() arg 1 must be a class

/path/to/env/lib/python3.10/abc.py:123: TypeError
=========================== short test summary info ============================
FAILED sktime/registry/tests/test_lookup.py::test_all_estimators_by_scitype[estimator_scitype20-True]
====================== 1 failed, 137 deselected in 1.46s =======================
```

fkiraly commented:
Random guess - could the issue be a decorated class? Where the outer object is a decorator rather than a class.

yarnabrina commented:
I do not know what is a decorated class, definitely decorators can be a reason. The `klass` seems to have type of an argument instead of the type of an estimator, so something weird must be happening somewhere.

yarnabrina commented:
@fkiraly I have rebased my branch from `main` and additionally removed use of `abc.ABC`. This is a subset of changes you added in #6537 and on my local is enough to pass lookup tests. I still think the reason is something else, but can't spot it and don't have enough bandwidth.

If you push any commits here, requesting you to delete your local branch first and create a new one (since I changed commit history with rebase).

yarnabrina commented:
@fkiraly @benHeid The CI too long in first run with failures only from cancelled jobs due to timeouts, and second run is still ongoing. But my impression is we are good to merge this before the current release.

Given the fact that it affects quite a few files, even though changes are extremely minimal, there can be a consideration to not merge it in patch release `0.30.2` and only do in minor `0.31.0`. That may lead to conflicts if some other PR's introduce old style codes, so my suggestion is to merge in patch release only, if you both agree.

fkiraly commented:
My plan was 0.31.0 due to large number of files changed, from the above I am not sure which of the options is your preference?

yarnabrina commented:
My preference is 0.30.2. I'm almost certain of major conflicts happening otherwise, and too much changes mean very long (apparently >1 day overall - not per job) CI run time.
